crates/matrix-sdk/src/widget_api/capabilities.rs

```diff
@@ -1,70 +1,50 @@
 use async_trait::async_trait;
+use tokio::sync::mpsc::Receiver;
 
 use super::{
     messages::{
-        capabilities::{EventFilter, Options},
-        MatrixEvent, from_widget::{SendEventResponse, SendToDeviceRequest},
+        capabilities::Options,
+        from_widget::{SendEventResponse, SendToDeviceRequest, ReadEventRequest, SendEventRequest, ReadEventResponse},
+        MatrixEvent,
     },
     Result,
 };
 
 /// A wrapper for the matrix client that only exposes what is available through the capabilities.
 #[allow(missing_debug_implementations)]
 pub struct Capabilities {
-    pub event_reader: Option<Box<dyn EventReader>>,
-    pub event_writer: Option<Box<dyn EventWriter>>,
+    // The options are a private member,
+    // they contain the type filters for the listeners.
+    options: Options,
```

Comment:
I think `options` should not be part of this structure, or, more precisely, I think this particular structure (`Capabilities`) should probably part of the `handler` module as it once was as it's definitely how `handler` defines the "expectations" from the driver that can implement capabilities. Then, the `capabilities` module may rather describe something more like an implementation of a module that constructs the actual `capabilities` that are then used by the handler.

Reply:
I don't have problems with `Options` stored as such, but rather my concern was that `Capabilities` is something that the `handler` defines as its requirements to the implementer, i.e. "hey, I expect that once you manage to give me capabilities, these must be there so that I can work with them". If we make separate fields for each capability, then we could always transform one into another.

Reply:
Maybe let me give you some examples to explain what I meant.

**Example 1:** If we know `X` and we can calculate `Y` based on `X` and such a calculation is not expensive, then instead of storing both `X` and `Y` (and keeping them in sync), we provide `f: X -> Y`. In our case, `X` are the `Capabilities` (the actual ones that the widget/driver granted), `Y` are the options that correspond to the capabilities that were granted, and `f: X -> Y` is the `From::from()` function. But if we store both in a structure, then it's like having both `X` and `Y` stored (meaning that they must be synchronized) and we don't have `f: X -> Y` (the `From::from` becomes a "getter", instead of becoming a conversion function).

**Example 2:** We have 3 objects (`a`, `b`, `c`) in a structure. But sometimes we need an array that contains these 3 objects. One way to solve it is to let the structure contain `a, `b`, `c`, `vec![a, b, c]`. This way we have 4 fields, one of which is a combination of the former 3. This works, but it has a problem that someone must keep them in sync, i.e. ensuring that the forth field is always a sum of 3 former fields. Another solution is to store only 3 original fields + define a function that returns/constructs `vec![a, b, c]` when it's called. This way, we don't have a problem with the need to keep in sync the forth field with others, but instead, we always construct the vector based on the actual things stored in the structure on-demand. Generally, the later is always preferred and much more robust/reliable way to write things, except for the cases when such a calculation/function is expensive (memory/performance-wise), in which case the first solution is taken (but then it means that you're essentially implementing some sort of cache).

Reply:
In this case the filter array are not really derivable they always will be more or less a getter.
You are right about the other boolean flags however. Here i totally see how it is more elegant to compute them from the capabilities Optional fields.

My intial idea also was to store each filter in a seperate field. and than compose a new Options struct in the from method and copy over the filters and compute the boolean flags.
But in the end that would have been a LOT more code 10x approximatly.
My thought process was, that Options can not be changed inside the Capabilites if we make sure that all methods are non mut. Than the options will always stay in sync and the only way to change them is to create new Capabilites

Reply:
I can of course also change it to storing the filters inside the capabilites and computing the other filds for the Options

Reply:
> My thought process was, that Options can not be changed inside the Capabilites if we make sure that all methods are non mut.

Yet there are 2 things to keep in mind:
1. The creator of actual `Capabilities` will still need to construct `Option`s on top of the capabilities that are actually granted to make sure that they match.
2. The one who works with capabilities can mutate one of the fields which would mean that internal `Options` would not represent the actual capabilities. I.e. if I as a receiver do something stupid like replacing one of the public fields with something else, I would assume that `From::from()` constructs `Option`s that **correspond** to the meaning of other fields, so it would be a surprising thing that changing certain fields in the structure construct the set of `Options` that don't correspond to what's inside the structure.

Note, that storing `options` inside the structure does not really eliminate code in the long run, because the code to create `options` that correspond to the actual capabilities is something that the creator of `Capabilities` would need to care about, so by storing it inside we don't really simplify things.

Reply:
So basically in the initialize function there could be a bug. options{some_cap: false} but still `Capabilites.some_cap = Some(some_cap_executor)` is not `None`. Now we it would make things less broken if retriving the capabilites at least would represent what the current capabilites are. Even though it does not match what we expected them to be.